src/dune_engine/sandbox.ml

```diff
@@ -90,15 +103,22 @@ let link_function ~(mode : Sandbox_mode.some) =
         | false -> fun src dst -> Io.portable_symlink ~src ~dst)
      | Copy ->
        fun src dst ->
-         let { Unix.st_kind; _ } = Path.Untracked.stat_exn src in
-         copy_recursively st_kind ~src ~dst
+         let what =
+           let { Unix.st_kind; st_perm; _ } = Path.Untracked.stat_exn src in
+           match st_kind with
+           | S_DIR -> `Dir st_perm
+           | S_REG -> `File
+           | kind -> `Other kind
+         in
+         copy_recursively what ~src ~dst
      | Hardlink ->
        (match Sys.win32 with
         | true -> win32_error mode
         | false -> fun src dst -> Io.portable_hardlink ~src ~dst)
      | Patch_back_source_tree ->
        (* We need to let the action modify its dependencies, so we copy
           dependencies and make them writable. *)
+       (* CR-someday rgrinberg: this doesn't work with directory targets *)
```

Comment:
I'm confused: doesn't `copy_recursively` exist solely for supporting directory targets (or, more accurately, for supporting dependencies on them)? It seems odd to go through all the trouble and not support directory targets in the end.

Reply:
I think @emillon just forgot to add full support for them here. I have upcoming PR's that will do so though. This PR is just focusing on the permissions.